### ISSUE TITLE
[css-syntax-3] The `+` in `<an+b>` confuses parsing tools

### ISSUE THREAD

**Issue by tidoust** [w3c/csswg-drafts#9473]
The [`<an+b>` type](https://drafts.csswg.org/css-syntax-3/#the-anb-type) is defined with a production rule that starts with `<an+b> = ...`. This type is now referenced from the definition of [`:nth()`](https://drafts.csswg.org/css-gcpm/#selectordef-nth) in the CSS Generated Content for Paged Media Module.

Problem is that the `+` confuses grammar parsers, including [CSSTree](https://github.com/csstree/csstree/issues/263) and the highlighting mechanism in Bikeshed (which creates a link to the definition of `+`), as they parse `+` as a [component value multiplier](https://drafts.csswg.org/css-values-3/#component-multipliers).

The [Value Syntax Definition](https://drafts.csswg.org/css-values-3/#component-types) is not entirely clear about what characters can be used to define basic data types between `<` and `>` but it says:
> Other non-keyword literal characters that may appear in a component value, such as “+”, must be written enclosed in single quotes.

I think that's the only occurrence in CSS specs of a construct that parsers find problematic. Tools could be fixed of course, I'm wondering whether it would be ok to defined the type as `<an-plus-b>` instead.

(If so, I'm happy to prepare a PR to update the specs that define or reference the type)

**Comment by cdoublev** [comment]
I think those tools should be able to parse any [`<ident>`](https://drafts.csswg.org/css-syntax-3/#ref-for-typedef-ident-token%E2%91%A0) between `<` and `>`, like in the [`syntax` descriptor](https://drafts.css-houdini.org/css-properties-values-api-1/#the-syntax-descriptor).

https://drafts.css-houdini.org/css-properties-values-api-1/#consume-a-data-type-name

**Comment by tabatkins** [comment]
I think it's fine to just rename it to `<an-b>`, honestly. The plus isn't necessary.

**Comment by cdoublev** [comment]
I do not see any mention in CSS GCPM 4: are you are thinking to `:nth-of-page(n)`? I am not sure if the spec authors expect `<integer>` or `<an-b>`. There are mentions in CSS GCPM 3 and CSS Overflow 4 though.

This PR might also be an opportunity to replace *"the An+B notation"* with *"the `<an-b>` type"*, and `An+B` with `<an-b>` in value definitions, to fix #9494.

**Comment by fantasai** [comment]
Question about bikeshedding: as we're getting further away from the actual form of the notation "2n+1", it's getting more confusing to refer to the notation. An+B was somewhat understandable, because it pattern-matches against the notation. But `<an-b>` just seems like weird 1970s computer garbage naming.

Some possible replacements:

* `<An-B>` at least helps it parse as three components (a, n, b), rather than two (an, b).
* `<ith>` or `<index>` or something like that
* other ideas?

**Comment by cdoublev** [comment]
> other ideas?

`<position-pattern>`.

I assume escaping `+` in `an+b`, which would allow to be a valid `<ident>` parsed by CSSTree, is excluded.

**Comment by tabatkins** [comment]
Yeah, we need to name that's useful to read and write; there's no reason to contort ourselves to preserve a specific name. ^_^

**Comment by Psychpsyo** [comment]
I like `<a-n-plus-b>` and `<nth>` the most. Maybe `<nth-notation>` so that it at least feels like a noun.
Anything that involves fully lowercase `an` reads like the word "an" which feels weird and a bit ugly to read to me.

`<nth>` or `<nth-notation>` would also acknowledge the fact that these can be of the form An-B as well. (or any other form that isn't strictly An+B)

**Comment by fantasai** [comment]
Re-opening because I think `<a-n-plus-b>` is just the most awkward option out of all of the ones suggested in this issue, and I'd like us to consider one of the less awkward ones. 😓

**Comment by tabatkins** [comment]
While I don't have a strong opinion on what we use, I don't think a-n-plus-b is particularly bad - it's exactly how we *pronounce* the construct already (like, we say "nth-child starts with an 'A N plus B' argument"). It's also not too long; if anything, it's on the shorter side character-wise. I think it's just striking you because, tho short, it's still *four words*, and one of them is us spelling out a character.

I think we should stick with the resolution. It's how we all pronounce `An+B` when actually talking about it, and it's always convenient when the production is the same term as the actual definition.

**Comment by kbabbitt** [comment]
+1 that I don't think `<a-n-plus-b>` is particularly bad. I think `<nth>` or `<nth-index>` would also be fine, but I'd prefer to avoid `<An-B>` or other forms that rely on casing to convey meaning.

**Comment by svgeesus** [comment]
> I think we should stick with the resolution. It's how we all pronounce `An+B` when actually talking about it,

I agree, it seems like a fine name to me for that reason.
